### ISSUE TITLE
item.Template.Fields randomly fails assertions for generated templates

### ISSUE THREAD

**Issue by pveller** [sergeyshushlyapin/Sitecore.FakeDb#100]
Not sure why. I pulled down latest FakeDb and added it there to reproduce. It does indeed fail. I am looking into it right now. It was working before (I had an older version in one of my projects and a test that was relying on these conditions to pass started failing after FakeDb upgrade). It also fails *randomly*. It may fail on first assertion, may fail on second. I tried re-arranging them and it would find the field by name but fail finding it by ID. Weird.

```csharp
    [Fact]
    public void ShouldProperlyCopyFieldsFromItemToTheTemplate()
    {
      // arrange
      var titleFieldId = ID.NewID;
      var descriptionFieldId = ID.NewID;

      // act
      using (var db = new Db
      {
        new DbItem("home")
        {
          new DbField("Title", titleFieldId) { Value = "Title" },
          new DbField("Description", descriptionFieldId) { Value = "Description" }
        }
      })
      {
        var home = db.GetItem("/sitecore/content/home");

        // assert
        home.Template.Fields.Count(f => f.Name == "Title").Should().Be(1);
        home.Template.Fields.Count(f => f.Name == "Description").Should().Be(1);
        home.Template.Fields.Count(f => f.ID == titleFieldId).Should().Be(1);
        home.Template.Fields.Count(f => f.ID == descriptionFieldId).Should().Be(1);
      }
    }
```

**Comment by sergeyshushlyapin** [comment]
Cool scenario! I found that it was broken after the #80 fix. Needs to be researched.

**Comment by pveller** [comment]
I was thinking about the same. I remember @mikeedwards83  asking about field items and I also saw that FakeDb now creates them when I was debugging this one. My plan was the same - start going back in time to see when it started failing. You saved me good 30 minutes :)

**Comment by pveller** [comment]
And while I am looking into it I might as well add support for `Source` (the field on the field item that goes through `getLookupSourceItem` pipeline). No objections I assume? We need it :) /cc @brianbeckham

**Comment by pveller** [comment]
There's something about `item.Template`. Here's an updated test. It still fails where it used to fail. New assertions are all passing OK. Apparently getting a template via `TemplateManager` does something different than getting it via `item.Template`:

```csharp
    [Fact]
    public void ShouldProperlyCopyFieldsFromItemToTheTemplate()
    {
      // arrange
      var titleFieldId = ID.NewID;
      var descriptionFieldId = ID.NewID;

      // act
      using (var db = new Db
      {
        new DbItem("home")
        {
          new DbField("Title", titleFieldId) { Value = "Title" },
          new DbField("Description", descriptionFieldId) { Value = "Description" }
        }
      })
      {
        var home = db.GetItem("/sitecore/content/home");

        var templateItem = db.GetItem("/sitecore/templates/home");
        templateItem.Children[0].Children.Count.Should().Be(2);

        var template = TemplateManager.GetTemplate(home.TemplateID, db.Database);
        template.GetFields(false).Count(f => f.Name == "Title").Should().Be(1);
        template.GetFields(false).Count(f => f.Name == "Description").Should().Be(1);

        // assert
        home.Template.Fields.Count(f => f.Name == "Title").Should().Be(1); // <-- Fails
        home.Template.Fields.Count(f => f.Name == "Description").Should().Be(1);
        home.Template.Fields.Count(f => f.ID == titleFieldId).Should().Be(1);
        home.Template.Fields.Count(f => f.ID == descriptionFieldId).Should().Be(1);
      }
    }
```

@sergeyshushlyapin , rings a bell maybe?

**Comment by pveller** [comment]
The problem seems to originate from `TemplateItem.GetTemplateFieldItem()`. This is where `template.GetFields()` convert into `TemplateFieldItem[]`. Too many different APIs to represent a template I must say :)

**Comment by pveller** [comment]
Here's where I am right now in my quest. Trying to recreate what Sitecore does to understand what we're missing:

```csharp
      var template = TemplateManager.GetTemplate(home.TemplateID, db.Database);
      var templateItem = new TemplateItem(db.Database.GetItem(home.TemplateID, home.Language));

      foreach (var field in template.GetFields(true))
      {
        TemplateSectionItem section = templateItem.GetSection(field.Section.ID);
        if (section == null && !field.Name.StartsWith("__"))
        {
          throw new Exception(field.Name); // <-- fails here for Title
        }

        Item innerItem = db.Database.Items[field.ID, templateItem.InnerItem.Language];
        if (innerItem == null && !field.Name.StartsWith("__"))
        {
          throw new Exception(field.Name);
        }
      }
```

**Comment by pveller** [comment]
I found it I believe. Working on a fix now. Basically, `TemplateTreeBuilder` creates a `Data` section as a child item of a template with an auto-generated ID. Then, our implementation of the `GetTemplates` on the `FakeDbDataProvider` will create a template with a `Data` section that is again given a new ID. This is the problem. This new ID is not known to FakeDb as an item and later attempts to get a section item will return `null`.

I will make it so that we only generate new `Data` section when a template didn't have one. @sergeyshushlyapin , pull request with a fix is right around the corner.

**Comment by sergeyshushlyapin** [comment]
Yes! That should be the case. Shouldn't you always generate a Data section **item** with the known section ID in the `TemplateTreeBuilder`? Otherwise #80 won't work I think..

**Comment by sergeyshushlyapin** [comment]
Wait... there is no *known* section id...

**Comment by pveller** [comment]
I haven't changed the `TemplateTreeBuilder`. I changed `FakeDbDataProvider` to reuse the `Data` section's ID if one can be found (i.e. we use what we have generated before for #80)

**Comment by pveller** [comment]
@sergeyshushlyapin, take a look at #101. Seams to be very straightforward and harmless. My test is now passing and everything else is still passing as well, of course :) Will now work on the `Source` support. Would be nice to see a release early next week. Thanks!

**Comment by sergeyshushlyapin** [comment]
Thank you! Looking on it right now.

**Comment by sergeyshushlyapin** [comment]
Published as v0.31.6. Thanks for your efforts, @pveller!
